xlators/cluster/dht/src/dht-layout.c

```diff
@@ -56,41 +56,67 @@ dht_layout_get(xlator_t *this, inode_t *inode)
 {
     dht_layout_t *layout = NULL;
     int ret = 0;
+    uint64_t ctx_int = 0;
+    dht_inode_ctx_t *ctx = NULL;
 
-    ret = dht_inode_ctx_layout_get(inode, this, &layout);
-    if ((!ret) && layout) {
-        GF_ATOMIC_INC(layout->ref);
+    LOCK(&inode->lock);
+    {
+        ret = __inode_ctx_get(inode, this, &ctx_int);
+        if (!ret) {
+            ctx = (dht_inode_ctx_t *)(uintptr_t)ctx_int;
+            if (ctx && ctx->layout) {
+                layout = ctx->layout;
+                GF_ATOMIC_INC(layout->ref);
+            }
+        }
     }
+    UNLOCK(&inode->lock);
+
     return layout;
 }
 
 int
 dht_layout_set(xlator_t *this, inode_t *inode, dht_layout_t *layout)
 {
     dht_conf_t *conf = NULL;
-    int oldret = -1;
     int ret = -1;
-    dht_layout_t *old_layout;
+    uint64_t ctx_int = 0;
+    dht_inode_ctx_t *ctx = NULL;
 
     conf = this->private;
     if (!conf || !layout)
         goto out;
 
-    LOCK(&conf->layout_lock);
+    LOCK(&inode->lock);
     {
-        oldret = dht_inode_ctx_layout_get(inode, this, &old_layout);
+        ret = __inode_ctx_get(inode, this, &ctx_int);
+        if (!ret) {
+            ctx = (dht_inode_ctx_t *)(uintptr_t)ctx_int;
+            if (ctx && ctx->layout) {
+                dht_layout_unref(ctx->layout);
+            }
+        }
+
         if (layout)
             GF_ATOMIC_INC(layout->ref);
```

Comment:
I would delay the refcount increment until we are sure that everything succeeds. This way we avoid 2 atomic operations in case of error.

Reply:
done